File: apps/csharp/design-pattern/src/DesignPattern/Patterns/TemplateMethod/Report.cs
namespace DesignPattern.Patterns.TemplateMethod;

public abstract class Report
{
    public string Title { get; }
    public string[] Text { get; }

    protected Report(string title, string[] text)
    {
        Title = title;
        Text = text;
    }

    public string OutputReport()
    {
        var result = OutputStart();
        result += OutputHead();
        result += OutputBodyStart();
        result += OutputBody();
        result += OutputBodyEnd();
        result += OutputEnd();
        return result;
    }

    protected virtual string OutputStart() => "";
    protected virtual string OutputHead() => $"  {Title}\n";
    protected virtual string OutputBodyStart() => "";

    protected string OutputBody()
    {
        var result = "";
        foreach (var line in Text)
        {
            result += OutputLine(line);
        }
        return result;
    }

    protected abstract string OutputLine(string line);
    protected virtual string OutputBodyEnd() => "";
    protected virtual string OutputEnd() => "";
}

public class HtmlReport : Report
{
    public HtmlReport(string title, string[] text) : base(title, text) { }

    protected override string OutputStart() => "<html>\n";
    protected override string OutputHead() => $"  <head><title>{Title}</title></head>\n";
    protected override string OutputBodyStart() => "  <body>\n";
    protected override string OutputLine(string line) => $"    <p>{line}</p>\n";
    protected override string OutputBodyEnd() => "  </body>\n";
    protected override string OutputEnd() => "</html>\n";
}

public class PlainTextReport : Report
{
    public PlainTextReport(string title, string[] text) : base(title, text) { }

    protected override string OutputHead() => $"***** {Title} *****\n";
    protected override string OutputLine(string line) => $"{line}\n";
}

File: apps/ruby/design-pattern/04-template-method/lib/report.rb
module TemplateMethod
  class Report
    def initialize
      @title = "月次報告"
      @text = ["順調", "最高の調子"]
    end

    def output_report
      output_start
      output_head
      output_body_start
      output_body
      output_body_end
      output_end
    end

    def output_body
      @text.each { |line| output_line(line) }
    end

    def output_start; end

    def output_head
      output_line(@title)
    end

    def output_body_start; end

    def output_line(_line)
      raise NotImplementedError, "サブクラスで output_line を実装してください"
    end

    def output_body_end; end

    def output_end; end
  end
end
